### PR TITLE
Add option to filter exporter endpoints

### PR DESCRIPTION
**Pull request by glrf**
This PR adds the basis  for the exporter-filterproxy.

### How this should work in practice

The `exporter-filterproxy` can proxy one or more Prometheus exporters. For example it might expose the kube-state-metrics exporter under `/kube-state-metrics`. Further it can filter the metrics. So if we call `/kube-state-metrics?namespace=foobar` it will only return metrics for the namespace `foobar`.

So for our usecase, we deploy the filterproxy once per cluster and each HAProxy of our services will forward requests for `<namespace>-cluster.dbaas.example.com/kube-state-metrics` to `filterproxy.svc/kube-state-metrics?namespace=<namespace>`. That way we only expose the metrics of that namespace.

### How it's implemented

There are three interesting parts to this:

`fetch.go`: Fetches the upstream metrics and parses them using a Prometheus package. I implemented some cache here as we had some concerns about scalability and load. With caching my laptop could easily handle ~8000 req/s while only sending a request to the upstream exporter every 5 seconds.

`filter.go`: Does the actual filtering of metrics. Looks at every metric and checks that it contains all specified labels.

`handler.go`: The actual http handler that calls `fetch` and `filter`.

### What's missing
The following is missing and will be added in a separate PR once I have an MVP running on the test cluster

* Exposing "dynamic" exporters such as the kubelet. This is currently not possible as there is no fixed address and the number of kubelets varies. We can use the headless service to discover these.
* Make service discovery less painful by implementing [http service discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#http_sd_config). This will be more important once we can expose the kubelet and we have a dynamic number of exposed endpoints.

## Checklist

- [x] The PR has a meaningful title. It will be used to auto generate the
      changelog.
      The PR has a meaningful description that sums up the change. It will be
      linked in the changelog.
- [x] PR contains a single logical change (to build a better changelog).
- [x] Update the documentation.
- [x] Categorize the PR by adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+type endpointConfig struct {
+	Path               string        `yaml:"path"`
+	Target             string        `yaml:"target"`
+	RefreshInterval    time.Duration `yaml:"refresh_interval"`
+	Auth               endpointAuth  `yaml:"auth"`
+	InsecureSkipVerify bool          `yaml:"insecure_skip_verify"`
+}
+
+type endpointAuth struct {
+	Type  authType `yaml:"type"`
```
Comment (bastjan) [inline config.go:24]: I'm not 100% sure yaml respects go "enums".

**Review by bastjan — APPROVED** [review]
Looks good! 😊 (also awesome PR description 😉)

File: config.go (hunk 1 above)
Comment (glrf) [reply config.go:24]: As far as I can tell it just treats it as a string while parsing and works fine. Of course it won't make sure that it's one of the predefined authTypes but I we catch that in main

Comment (bastjan) [reply config.go:24]: I missed that it's checked in main 😅

**Review by zugao — APPROVED** [review]
LGTM. Can you add a diagram and show how all things are connected?

**Comment by glrf** [comment]
> LGTM. Can you add a diagram and show how all things are connected?

Is it ok for you if I do that once I implemented the service discovery for the kubelet?

**Comment by zugao** [comment]
> > LGTM. Can you add a diagram and show how all things are connected?
> 
> Is it ok for you if I do that once I implemented the service discovery for the kubelet?

Absolutely!

**Pull request merged**